File: src/main/java/fi/quanfoxes/parser/nodes/UnresolvedIdentifier.java
package fi.quanfoxes.parser.nodes;

import fi.quanfoxes.parser.Context;
import fi.quanfoxes.parser.Node;
import fi.quanfoxes.parser.Resolvable;

public class UnresolvedIdentifier extends Node implements Resolvable {
    private String value;

    public UnresolvedIdentifier(String value) {
        this.value = value;
    }

    public Node getResolvedNode(Context context) throws Exception {
        if (context.isTypeDeclared(value)) {
            return new TypeNode(context.getType(value));
        }
        else if (context.isVariableDeclared(value)) {
            return new VariableNode(context.getVariable(value));
        }
        else {
            throw new Exception(String.format("Couldn't resolve identifier '%s'", value));
        }
    }

    @Override
    public Node resolve(Context context) throws Exception {
        return getResolvedNode(context);
    }

    @Override
    public NodeType getNodeType() {
        return NodeType.UNRESOLVED_IDENTIFIER;
    }
}

File: Zigzag/Parser/Nodes/UnresolvedIdentifier.cs
using System;

public class UnresolvedIdentifier : Node, IResolvable, IType
{
	public string Value { get; private set; }

	public UnresolvedIdentifier(string value)
	{
		Value = value;
	}

	public Node? GetResolvedNode(Context context)
	{
		return Singleton.GetIdentifier(context, new IdentifierToken(Value), Parent?.Is(NodeType.LINK_NODE) ?? false);
	}

	public Node? Resolve(Context context)
	{
		return GetResolvedNode(context);
	}

	public new Type? GetType()
	{
		return Types.UNKNOWN;
	}

	public override NodeType GetNodeType()
	{
		return NodeType.UNRESOLVED_IDENTIFIER;
	}

	public Status GetStatus()
	{
		return Status.Error($"Couldn't resolve identifier '{Value}'");
	}

	public override bool Equals(object? obj)
	{
		return obj is UnresolvedIdentifier identifier &&
				base.Equals(obj) &&
				Value == identifier.Value;
	}

	public override int GetHashCode()
	{
		return HashCode.Combine(base.GetHashCode(), Value);
	}
}